### PR TITLE
Make sure systems are up to date when they're output

### LINKED ISSUE #4474: Scalar Variables Not Updated Before Output
Certain circumstances can lead to ScalarVariables not having their most up to date value when they are output.  I'll put in a test and a fix for this.

This is critical because @mellis13 needs it immediately.

### PR DESCRIPTION
**Pull request by friedmud**
Fixes an issue @mellis13 uncovered where nl_sys and aux_sys might not get `update()` called on them before being output.

To fix it I changed the Transient Executioner to go through FEProblem for outputting instead of calling the OutputWarehouse directly.  This allow FEProblem a chance to update the nl and aux systems first...

closes #4474

### CODE REVIEW

File: framework/src/base/FEProblem.C [hunk 1]
```diff
@@ -3129,6 +3130,16 @@ void FEProblem::restoreSolutions()
     _displaced_problem->updateMesh(*_nl.currentSolution(), *_aux.currentSolution());
 }
 
+void
+FEProblem::outputStep(ExecFlagType type)
```
Comment (aeslaughter) [inline framework/src/base/FEProblem.C]: This makes me very sad, I tried very hard to keep FEProblem out of the output mix. I would really like all output related stuff go through the warehouse. If we do this, we need to move other functions as well to keep the API consistent. I don't want to see calls to both the OutputWarehouse and FEProblem from things, it should be one or the other.

**Comment by aeslaughter** [comment]
I am against this, I would rather see the OutputWarehouse have pointers to the Aux and NL systems than add another layer of calls to FEProblem. As it is now, all calls that produce output go through the OutputWarehouse, which I think is consistent and logical. I would like to see more MOOSE objects have this behavior.

**Comment by friedmud** [comment]
The flow of that logic and dependencies is backwards though.  What if you have another kind of `Problem` that doesn't have Aux or NL (like a `Problem` that runs a 3rd party code).  What about the displaced stuff?

Leaking the internal details of what is going on inside of Problem is bad.  @andrsd initially wrote the Problem class for just this purpose: to shield other systems from needing to juggle NL, Aux and Displaced so they could be more general.

That's why we call `FEProvlem::reinit()` instead of `NL::reinit()`,  `Aux::reinit()` etc directly.

Note: I don't feel too strongly about this and could definitely have my mind changed... I'm just discussing the options here...

**Comment by friedmud** [comment]
Another thing: what if you have a special kind of Problem that needs to copy some data around just before output?  Like syncing with an external code or something?

**Comment by aeslaughter** [comment]
I see your point, but I think that Problem should have an update() that does the necessary updates then, just like FEProblem::reinit(), as mentioned. This way OutputWarehouse can have a pointer to FEProblem and call update() and it should do the correct thing, regardless if people create some crazy Problem object.

**Comment by friedmud** [comment]
I'm not against that.

One more then: (theoretically) what if you have something you are outputting that has nothing to do with a `Problem` at all... as in you DONT want to `update` the Problem at that time?

**Comment by aeslaughter** [comment]
Use a flag to disable update??? outputStep(ExecFlagType type, problem_update=true);

**Comment by friedmud** [comment]
Should be ok.  What do you think @permcody ?

**Comment by permcody** [comment]
Still doesn't seem as clean as it could be but my head isn't wrapped around this problem like the two of you have been doing. Any solution that separates the duties of these two systems without built-in assumptions will be best for future expansion. If this solves your problem and Andrew is onboard then go ahead and merge it in.

**Comment by friedmud** [comment]
Yeah - this might not be the perfect solution... but it should work for now.  It still feels kinda wrong to have the Output system calling functions on FEProblem...

**Comment by aeslaughter** [comment]
I am ok with going with the FEProblem approach, but I think a few more methods might be needed so that the OutputwareHouse is not used by executioners or anything else, which maybe you did, I didn't go through everything. I would also want to make OutputWarehouse::outputStep() private and friend FEProblem, so that people don't try to call the wrong one. The apps will also need to be updated.

**Comment by aeslaughter** [comment]
@friedmud I can take this over to get everything moved to FEProblem.

**Comment by friedmud** [comment]
Sounds good - are you going to build on what I've done here - or start over?

**Comment by aeslaughter** [comment]
I will start from you branch and submit a PR back to you.

**Comment by friedmud** [comment]
Sounds good - thanks!

**Comment by aeslaughter** [comment]
@friedmud Your changes to `framework/src/transfers/MultiAppPostprocessorToAuxScalarTransfer.C` seem to be causing a Exodiff in parallel, could you take a look? If not, I will dig in.

**Comment by friedmud** [comment]
I'll take a look at it...

**Comment by friedmud** [comment]
Ok - I found the problem... it was a logic error on my part.  I was trying to use the ScalarVariable interfaces... but in this case you can't... tricky...

**Comment by friedmud** [comment]
If this passes the tests it's ready to go....

**Comment by friedmud** [comment]
Interesting... is this an "integration" problem?  Like I need to rebase this branch?

**Comment by aeslaughter** [comment]
@friedmud I just tested your PR, it compiled and passed all modules tests, something strange is happening.

**Comment by friedmud** [comment]
I think this is ready...

**Comment by aeslaughter** [comment]
I will merge this later today, I need to double check my application patches now that they have been sitting around for awhile. I also think Hyrax was moved, so I need to move that patch from svn.

**Pull request merged**